Source/x11/XGServerWindow.m

```diff
@@ -4409,75 +4415,140 @@ - (void) freecursor: (void*) cid
   return depth;
 }
 
+/* This method assumes that we deal with one X11 screen - `defScreen`.
+   Basically it means that we have DISPLAY variable set to `:0.0`.
+   AppKit and X11 use the same term "screen" with different meaning:
+   AppKit Screen - single monitor/display device.
+   X11 Screen - it's a plane where X Server can draw on.
+   XRandR - Xorg extension that helps to manipulate monitor layout providing
+   X11 Screen. 
+   We map XRandR monitors (outputs) to NSScreen. */
 - (NSArray *)screenList
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSArray        *screens;
+  NSMutableArray *tmpScreens;
+  int            i;
+
+  monitorsCount = 0;
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+  }
+  
+#ifdef HAVE_XRANDR
+  XRRScreenResources  *screen_res;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  if (screen_res != NULL)
+    {
+      monitorsCount = screen_res->noutput;
+      tmpScreens = [NSMutableArray arrayWithCapacity: monitorsCount];
+      monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+      for (i = 0; i < monitorsCount; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          monitors[i].screen_id = defScreen;
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+
+              // Fill the cache of device parameters
+              monitors[i].depth = [self windowDepthForScreen: defScreen];
+              monitors[i].resolution = [self resolutionForScreen: defScreen];
+              monitors[i].frame = NSMakeRect(crtc_info->x, crtc_info->y,
+                                             crtc_info->width, crtc_info->height);
+              // Add number of device
+              [tmpScreens addObject: [NSNumber numberWithInt: i]];
```

Comment:
That way we get a screen with the id 0, I am not sure whether our code will handle this correctly.

Reply:
I see no problem here. Our code handles it correctly.